runtime/fundamentals/testing.md

```diff
@@ -0,0 +1,757 @@
+
+To define a test in Deno, you use the `Deno.test()` function. Here are some
+examples:
+
+```ts title="my_test.ts"
+import { assertEquals } from "@std/assert";
+
+// basic test
+Deno.test("simple test", () => {
+  const x = 1 + 2;
+  assertEquals(x, 3);
+});
+
+// a named test
+Deno.test("addition test", () => {
+  const sum = 2 + 3;
+  assertEquals(sum, 5);
+});
+
+// async test:
+import { delay } from "@std/async";
+
+Deno.test("async test", async () => {
+  const x = 1 + 2;
+  await delay(100);
+  assertEquals(x, 3);
+});
+
+// test with permissions
+Deno.test({
+  name: "read file test",
+  permissions: { read: true },
+  fn: () => {
+    const data = Deno.readTextFileSync("./somefile.txt");
+    assertEquals(data, "expected content");
+  },
+});
+```
+
+If you prefer a "jest-like" `expect` style, the Deno standard library provides
+an `expect` function that can be used in place of `assertEquals`:
+
+```ts title="my_test.ts"
+import { expect } from "jsr:@std/expect";
+import { add } from "./add.js";
+
+Deno.test("add function adds two numbers correctly", () => {
+  const result = add(2, 3);
+  expect(result).toBe(5);
+});
+
+Deno.test("add function handles negative numbers", () => {
+  const result = add(-2, -3);
+  expect(result).toBe(-5);
+});
+```
+
+## Running Tests
+
+To run your tests, use the [`deno test`](/runtime/manual/tools/test/)
+subcommand.
+
+If run without a file name or directory name, this subcommand will automatically
+find and execute all tests in the current directory (recursively) that match the
+glob `{*_,*.,}test.{ts, tsx, mts, js, mjs, jsx}`.
+
+```sh
+# Run all tests in the current directory and all sub-directories
+deno test
+
+# Run all tests in the util directory
+deno test util/
+
+# Run just my_test.ts
+deno test my_test.ts
+
+# Run test modules in parallel
+deno test --parallel
+
+# Pass additional arguments to the test file
+deno test my_test.ts -- -e --foo --bar
+```
+
+## Test Steps
+
+Deno also supports test steps, which allow you to break down tests into smaller,
+manageable parts. This is useful for setup and teardown operations within a
+test:
+
+```ts
+Deno.test("database operations", async (t) => {
+  await t.step("insert user", async () => {
+    // Insert user logic
+  });
+  await t.step("insert book", async () => {
+    // Insert book logic
+  });
+});
+```
+
+## Command line filtering
+
+Deno allows you to run specific tests or groups of tests using the `--filter`
+option on the command line. This option accepts either a string or a pattern to
+match test names.
+
+Consider the following tests:
+
+```ts
+Deno.test({ name: "my-test", fn: myTest });
+Deno.test({ name: "test-1", fn: test1 });
+Deno.test({ name: "test-2", fn: test2 });
+```
+
+### Filtering by string
+
+To run all tests that contain the word "test" in their names, use:
+
+```sh
+deno test --filter "test" tests/
+```
+
+This command will execute `my-test`, `test-1`, and `test-2` because they all
+include "test" in their names.
```

Comment:
I feel like we should change the example because after filtering we still run _all_ tests. We should have an example for filtering where we only run a subset of the tests, not all of them.

Reply:
Good point